### PR TITLE
Fix use of non-invariant culture APIs

### LINKED ISSUE #4371: Wrong dates with changed culture
In the previous versions of linq2db, CurrentCulture didn't affect datetime values ​​stored in the database. For example, when we worked with Persian calendar dates in our web application, linq2db saved them (datetimes) as a Gregorian calendar values (InvariantCulture) to databases (not matter of the current culture).

But in the latest version of linq2db (version 5.3.2) this behaviour changed and now it causes issues. Now linq2db started saving dates in current culture. So if your web application is multilingual (hence, current culture could be different for diffrent HTTP requests), then dates are saved in different cultures. Please find more information on how reproduces the issue at  https://github.com/nopSolutions/nopCommerce/issues/7002

**Environment details**

`Linq To DB` version: *5.3.2*
Database: *Any*

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #4371

Found issues:
- DateTime[Offset].ToString call with exact format without format provider generates date using current calendar which produce date differ from gregorian calendar or crash if date cannot be used with such calendar
- LinqService serializer throws exception due to serialization dictionary conflicts. Caused by negative numers formatted differently in tested culture

### CODE REVIEW

**Review by igor-tkachev — DISMISSED** [review]

**Review by igor-tkachev — DISMISSED** [review]

File: Source/LinqToDB/Data/DataConnection.cs [hunk 1]
```diff
@@ -508,16 +510,16 @@ static void OnTraceInternal(TraceInfo info)
 				{
 					using var sb = Pools.StringBuilder.Allocate();
 
-					sb.Value.Append(info.TraceInfoStep);
+					sb.Value.AppendFormat(CultureInfo.InvariantCulture, "{0}", info.TraceInfoStep);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/DataConnection.cs:513]:
```suggestion
					sb.Value.Append(CultureInfo.InvariantCulture, $"{info.TraceInfoStep}");
```

Comment (viceroypenguin) [reply Source/LinqToDB/Data/DataConnection.cs:513]: And others in file

File: Source/LinqToDB/DataProvider/Access/AccessMappingSchema.cs [hunk 2]
```diff
@@ -54,11 +54,7 @@ static void ConvertBinaryToSql(StringBuilder stringBuilder, byte[] value)
 
 		static void AppendConversion(StringBuilder stringBuilder, int value)
 		{
-			stringBuilder
-				.Append("chr(")
-				.Append(value)
-				.Append(')')
-				;
+			stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "chr({0})", value);
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessMappingSchema.cs:57]:
```suggestion
			stringBuilder.Append(CultureInfo.InvariantCulture, "chr({value})");
```

File: Source/LinqToDB/DataProvider/Access/AccessODBCSchemaProvider.cs [hunk 3]
```diff
@@ -1,9 +1,12 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using System.Data;
 using System.Linq;
 
 namespace LinqToDB.DataProvider.Access
 {
+	using System.Globalization;
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessODBCSchemaProvider.cs:8]: Should be with other `System.*` usings at top

File: Source/LinqToDB/DataProvider/Access/AccessOleDbSchemaProvider.cs [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
 namespace LinqToDB.DataProvider.Access
 {
+	using System.Globalization;
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessOleDbSchemaProvider.cs:9]: Should be with other `System.*` usings at top

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseMappingSchema.cs [hunk 5]
```diff
@@ -221,7 +221,7 @@ private static void ConvertByteArray(StringBuilder stringBuilder, SqlDataType dt
 						if (i > 0)
 							stringBuilder.Append(':');
 
-						stringBuilder.AppendFormat("{0:x2}{1:x2}", value[i], value[i + 1]);
+						stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:x2}{1:x2}", value[i], value[i + 1]);
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/ClickHouse/ClickHouseMappingSchema.cs:224]:
```suggestion
						stringBuilder.Append(CultureInfo.InvariantCulture, $"{value[i]:x2}{value[i + 1]:x2}");
```

**Review by viceroypenguin — CHANGES_REQUESTED** [review]
I'm not going to go through every file and point out `sb.Append(.., $"")` improvements, but these will generally be better than `.AppendFormat()`.

**Review by igor-tkachev — APPROVED** [review]

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 6]
```diff
@@ -5,13 +5,18 @@
...
 using System.Dynamic;
+using System.Globalization;
 using System.Linq;
 using System.Linq.Expressions;
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
 
+#if NETFRAMEWORK || NETCOREAPP3_1 || NETSTANDARD2_0 || NETSTANDARD2_1
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/CommandInfo.cs:16]: This is not repeated in other files. Necessary?

**Review by viceroypenguin — DISMISSED** [review]

File: Source/LinqToDB/Data/CommandInfo.cs (hunk 6 above)
Comment (MaceWindu) [reply Source/LinqToDB/Data/CommandInfo.cs:16]: it depends on used types in file, some need it only for polyfills

**Pull request merged**